yktrace/src/hwt.rs

```diff
@@ -0,0 +1,94 @@
+//! Hardware tracing via ykrustc.
+
+use super::{SirTrace, ThreadTracer, ThreadTracerImpl};
+use crate::{errors::InvalidTraceError, maptrace::MapTrace};
+use core::yk::SirLoc;
+use hwtracer::backends::TracerBuilder;
+
+/// A trace collected via hardware tracing.
+#[derive(Debug)]
+struct HWTSirTrace {
+    sirtrace: Vec<SirLoc>
+}
+
+impl SirTrace for HWTSirTrace {
+    fn raw_len(&self) -> usize {
+        self.sirtrace.len()
+    }
+
+    fn raw_loc(&self, idx: usize) -> &SirLoc {
+        &self.sirtrace[idx]
+    }
+}
+
+/// Hardware thread tracer.
+struct HWTThreadTracer {
+    ttracer: Box<dyn hwtracer::ThreadTracer>
+}
+
+impl ThreadTracerImpl for HWTThreadTracer {
+    #[trace_tail]
+    fn stop_tracing(&mut self) -> Result<Box<dyn SirTrace>, InvalidTraceError> {
+        let hwtrace = self.ttracer.stop_tracing().unwrap();
+        let mt = MapTrace::new();
+        if let Some(sirtrace) = mt.map(hwtrace) {
+            Ok(Box::new(HWTSirTrace { sirtrace }) as Box<dyn SirTrace>)
+        } else {
+            Err(InvalidTraceError::InternalError)
+        }
+    }
+}
+
+#[trace_head]
+pub fn start_tracing() -> ThreadTracer {
+    let tracer = TracerBuilder::new().build().unwrap();
+    let mut ttracer = (*tracer).thread_tracer();
+    match ttracer.start_tracing() {
```

Comment:
I think you can make this bit shorter with `ttracer.start_tracing().unwrap_or_else(|e| panic!(...))`

Reply:
Or even `ttracer.start_tracing().expect("Failed to start tracer")` if the error implements `Debug`.